### PR TITLE
Enhance functionality of library

### PR DESCRIPTION
**Pull request by Apollon77**
Hey,

Because I needed it I enhanced some of the functionality. Beside some small jslint fixes I mainly implemented some other functions and streamlined the result parsing.

Tell me what you think and if you would accept that ... then I could also update the README. A feedback would be great because I need this functionality in my project.

### CODE REVIEW

**Comment by skarcha** [comment]
Hi @Apollon77 !

Oh! Thanks for your PR. I can't test it right now (problems with my USB connection) but it looks pretty good! ;)

I only see some mix of space/tab indentation. If you could fix them and update the README, I'd be glad to merge it. :-)

**Comment by Apollon77** [comment]
indentation: what you prefer? spaces or tabs?

**Comment by Apollon77** [comment]
One more: As you may see I added also parsing of result to messages like send command because it is possible that an error is returned. I added a callback with the error as parameter to return that to leave the on-error "event" for basic communication errors. Is that ok yor should it also send te error event here?

**Comment by skarcha** [comment]
I don't have a preference about indentation actually. I'm not a JS programmer. You can use what you want, but I think it'd be good if you use the indentation I used.

About the added callback, as I told you, I'm not a JS or NodeJS programmer so use what you think is better, but only if it doesn't broke the API. All NodeJS modules that use node-nut should works without modifications after this change. Do you agree?

**Comment by Apollon77** [comment]
sure, none of my changes should break anything so far

**Comment by skarcha** [comment]
When you update the README and think that it's fine, drop me a line and I'll merge it. THANKS! :-)

**Comment by Apollon77** [comment]
BTW: I added new error handling as "normal" in nodejs stuff to return a second parameter "err" to all callbacks to check if an error happened. Before errors were completely ignored for LIST returns and ended up is node-nut internal status being "waiting" forever. And on error the callback was never called at all, but this is not as awaited I think. So now callback is called but with an empty return, but a second error parameter.
It's a different behaviour, but should be not breaking at all ..
Alternative (to make sure) you could increase version to 1.0.0 and then it should be no conflict at all when dependencies are set as usual

And README updated too ... I will do final testing this evening and then drop you a line. maybe think on the version numbering querstion ? ;-))

**Comment by skarcha** [comment]
Ok, I think that the version increasing to 1.0.0 is the right way, so do all changes you consider are needed, even if they break backward compatibility. Then, please, test it all you can and drop me a line... ;-)

**Comment by Apollon77** [comment]
I tested everything and also the usage in my project was tested by a second person.

From my point of view it is green and ready for release :-)

Do you need me to change anything? maybe a "v1.0.0 vs before comatibility note" in README?

**Comment by skarcha** [comment]
Great! Yes, a note about the compatibility should be fine. Thanks @Apollon77 !

**Comment by Apollon77** [comment]
here you go :-)

File: README.md [hunk 1]
```diff
@@ -2,6 +2,13 @@
 
 Node-NUT is a NodeJS module that implements a NUT (Network UPS Tools) client.
 
+## Version compatibility information
+v1.0.0 introduce callbacks into all functions including a second "error" parameter for all gameaccounts. This error parameter is null on success or a text on error. Please see the updated method documentation below.
```
Comment (skarcha) [inline README.md]: What's a **gameaccout**? :-)

Comment (Apollon77) [reply README.md]: fixed :-)

**Comment by ToddGreenfield** [comment]
Hi. I have been working on a project using node-nut and just happened to check back only to see this PR, so I figured I would give it a try/test and installed from Apollon77's master branch. When I run the below test script (which works with skarcha's master) it now hangs after returning the list of UPS and does not return the list of UPSVars. I only have one ups connected, and I am also not a JS programmer... but something changed. I was interested in using the new on.error fix to capture if/when the nut server loses connection with a ups.

 var Nut = require('node-nut');
var _ = require('underscore');

oNut = new Nut(3493, 'localhost');

oNut.on('error', function(err) {
        console.log('There was an error: ' + err);
});

oNut.on('close', function() {
        console.log('Connection closed.');
});

oNut.on('ready', function() {
        self = this;
        var ups, upsFull;
        oNut.GetUPSList(function(upslist) {
                ups = _.keys(upslist);
                upsFull = upslist;
        })
        setTimeout(function() {
                console.log(upsFull);
                getVars(ups);
        }, 1000);
});

function getVars (ups) {
var upsvars;
        oNut.GetUPSVars(ups, function(upsvars) {
                console.log(upsvars);
                var obj = _.keys(upsvars);
                for (prop in obj) {
                        switch(obj[prop]) {
                                case "device.mfr":
                                        console.log("Nut eventVar device.mfr received is %s: ", upsvars["device.mfr"]);
                                        break;
                                case "device.model":
                                        console.log("Nut eventVar device.model received is %s: ", upsvars["device.model"]);
                                        break;
                                case "ups.productid":
                                        console.log("Nut eventVar ups.productid received is %s: ", upsvars["ups.productid"]);
                                        break;
                        }
                }
        self.close();
        return upsvars;
        })

}

oNut.start();

**Comment by skarcha** [comment]
Hi @ToddGreenfield, thanks for your feedback!. I hope @Apollon77 can test your code and give us more info.

**Comment by Apollon77** [comment]
I will chekc it this evening. But to make sure: The "on(error)" was also there before :-)

@ToddGreenfield : Can you also post the output you get from your script!?

**Comment by ToddGreenfield** [comment]
@Apollon77 Yes, on.error was there before but it is not able to detect if/when the nut client loses connection with the UPS, so I was hoping your PR fixed that... 
The output of my test script running on the old version (0.1.0) is:
pi@homepi-a:~/stage/homebridge-nut $ node test
[ mainups: 'Main UPS' ]
[ 'battery.charge': '100',
  'battery.charge.low': '10',
  'battery.charge.warning': '20',
  'battery.mfr.date': 'CPS',
  'battery.runtime': '5100',
  'battery.runtime.low': '300',
  'battery.type': 'PbAcid',
  'battery.voltage': '13.3',
  'battery.voltage.nominal': '12',
  'device.mfr': 'CPS',
  'device.model': 'UPS CP1000AVRLCD',
  'device.type': 'ups',
  'driver.name': 'usbhid-ups',
  'driver.parameter.pollfreq': '30',
  'driver.parameter.pollinterval': '2',
  'driver.parameter.port': 'auto',
  'driver.version': '2.7.2',
  'driver.version.data': 'CyberPower HID 0.3',
  'driver.version.internal': '0.38',
  'input.transfer.high': '140',
  'input.transfer.low': '90',
  'input.voltage': '119.0',
  'input.voltage.nominal': '120',
  'output.voltage': '119.0',
  'ups.beeper.status': 'enabled',
  'ups.delay.shutdown': '20',
  'ups.delay.start': '30',
  'ups.load': '6',
  'ups.mfr': 'CPS',
  'ups.model': 'UPS CP1000AVRLCD',
  'ups.productid': '0501',
  'ups.realpower.nominal': '600',
  'ups.status': 'OL',
  'ups.timer.shutdown': '-60',
  'ups.timer.start': '0',
  'ups.vendorid': '0764' ]
Nut eventVar device.mfr received is CPS:
Nut eventVar device.model received is UPS CP1000AVRLCD:
Nut eventVar ups.productid received is 0501:
Connection closed.

When run on the new version 1.0.0 the output is just:
pi@homepi-a:~/stage/homebridge-nut $ node test
[ mainups: 'Main UPS' ]

And hangs from there---
Thanks for checking into this...

**Comment by Apollon77** [comment]
The have imporoved error handling a bit further. Please test again.

I expect that you get an callback with an empty "upvars" object.
When you add the second "error" parameter to 

```
oNut.GetUPSVars(ups, function(upsvars) {
```

and output "error" I expect that you get "Old communication still running".

Please try that before doing the next:
In general your code is wrong in multiple places and do not wotk with the asyncronous style of javascript and will trigger exactly this problem. It is some kind of timing error.
1.) Your "setTimeout" to read the vars expects that the "getUpsList" is finished within 1000ms ... if not you start a second command while the first is still running and this was not handled well before
2.) Your code then do several getUpsVars at the same time, one per found UPS. but the commands need to be processed sequential and the communication only should be closed at the last one.

A correct code should more look like (hopefully, untested):

```
var Nut = require('node-nut');
var _ = require('underscore');

oNut = new Nut(3493, 'localhost');

oNut.on('error', function(err) {
    console.log('There was an error: ' + err);
});

oNut.on('close', function() {
    console.log('Connection closed.');
});

oNut.on('ready', function() {
    oNut.GetUPSList(function(upslist, err) {
        if (err) {
            console.log('ERROR ' + err);
        }
        else {
            var ups = _.keys(upslist);
            console.log(upslist);
            getVars(ups);
        }
    })
});

function getVars(ups) {
    if (!ups.length) {
        self.close();
        console.log('DONE');
    }
    var currentUps = ups.shift();
    oNut.GetUPSVars(currentUps, function(upsvars, err) {
        if (err) {
            console.log('ERROR ' + err);
            self.close();
        }
        else {
            console.log('UPS ' + currentUps + ' : ' + upsvars);
            var obj = _.keys(upsvars);
            for (prop in obj) {
                switch(obj[prop]) {
                    case "device.mfr":
                        console.log("Nut eventVar device.mfr received is %s: ", upsvars["device.mfr"]);
                        break;
                    case "device.model":
                        console.log("Nut eventVar device.model received is %s: ", upsvars["device.model"]);
                        break;
                    case "ups.productid":
                        console.log("Nut eventVar ups.productid received is %s: ", upsvars["ups.productid"]);
                        break;
                }
            }
            getVars(ups);
        }
    })
}

oNut.start();
```

**Comment by ToddGreenfield** [comment]
@Apollon77 First, thank you for your help here. I added err as the 2nd parm as suggested, and also did a cut/paste of the suggested code above and both tries resulted in the same - hanging after the initial list of UPS returned (which is only 1 in my case).

pi@homepi-a:~/stage/homebridge-nut $ node test3
[ mainups: 'Main UPS' ]
(hangs here.... until I cntrl-c)

**Comment by Apollon77** [comment]
did you also updated the used node-nut code with the latest changes from yesterday evening?

**Comment by Apollon77** [comment]
Ok, you were right, there were some glitches still in error handling ... now fixed hopefully. Pleas try.

And also here a (again and this time tested) correct version of your script:

```
var Nut = require('./node-nut');

oNut = new Nut(3493, 'cubietruck.fritz.box');

oNut.on('error', function(err) {
    console.log('There was an error: ' + err);
});

oNut.on('close', function() {
    console.log('Connection closed.');
});

oNut.on('ready', function() {
    oNut.GetUPSList(function(upslist, err) {
        if (err) {
            console.log('ERROR ' + err);
        }
        else {
            var ups = Object.keys(upslist);
            console.log(upslist);
            getVars(ups);
        }
    });
});

function getVars(ups) {
    if (!ups.length) {
        oNut.close();
        console.log('DONE');
        return;
    }
    var currentUps = ups.shift();
    oNut.GetUPSVars(currentUps, function(upsvars, err) {
        if (err) {
            console.log('ERROR ' + err);
            self.close();
        }
        else {
            console.log('UPS ' + currentUps + ' : ' + upsvars);
            var obj = Object.keys(upsvars);
            for (var prop in obj) {
                switch(obj[prop]) {
                    case "device.mfr":
                        console.log("Nut eventVar device.mfr received is %s: ", upsvars["device.mfr"]);
                        break;
                    case "device.model":
                        console.log("Nut eventVar device.model received is %s: ", upsvars["device.model"]);
                        break;
                    case "ups.productid":
                        console.log("Nut eventVar ups.productid received is %s: ", upsvars["ups.productid"]);
                        break;
                }
            }
            getVars(ups);
        }
    });
}

oNut.start();

```

**Comment by ToddGreenfield** [comment]
I did a quick test and this now works after 2 small test script changes- Changing the script back to using localhost, and using a separate line to send to console.log the upsvars (I believe since this is a type object and can't be combined with a string on the same output line).
I will test out my dependent package later today-
Thanks-

**Comment by skarcha** [comment]
Hi! Thanks guys for following up. As soon as you both test it and agree about the behaviour I'll merge it. Thanks again for you time.

**Comment by Apollon77** [comment]
my project runs with new version since fix this morning

**Comment by skarcha** [comment]
Hi @ToddGreenfield!

Did you test your dependent package?

**Comment by ToddGreenfield** [comment]
@skarcha Hi. Sorry for the delay. Yes, I just tested the full package (homebridge-nut) without any changes and it worked as expected. You can go ahead and merge the PR. Thanks again.

**Comment by Apollon77** [comment]
Cool! Thanks for a short info when you published 1.0.0 on npm so that I can change my library back from my Fork to npm :-)

**Comment by skarcha** [comment]
Published in npm! Thank you guys! In special to @Apollon77 for writing the code and @ToddGreenfield for the testing.

I have to test your projects as soon as I can... ;-)

**Pull request merged**